{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#data from\n",
    " https://brainmapportal-live-4cc80a57cd6e400d854-f7fdcae.divio-media.net/filer_public/94/2b/942bdfbc-89cb-4414-9eda-a348fadef841/mouse_patch-seq_demo.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"../src\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metadata file\n",
    "The metadata file has information about each Patch-seq cell, including its different identifiers and assigned cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>cell_specimen_id</th>\n",
       "      <th>cell_specimen_name</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>structure</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>donor_name</th>\n",
       "      <th>biological_sex</th>\n",
       "      <th>age</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>apical_dendrite_status</th>\n",
       "      <th>neuron_reconstruction_type</th>\n",
       "      <th>cell_soma_normalized_depth</th>\n",
       "      <th>depth_from_pia_um</th>\n",
       "      <th>ephys_session_id</th>\n",
       "      <th>transcriptomics_sample_id</th>\n",
       "      <th>transcriptomics_batch</th>\n",
       "      <th>T-type Accession ID</th>\n",
       "      <th>T-type Label</th>\n",
       "      <th>MET-type Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507</td>\n",
       "      <td>Vip-IRES-Cre;Ai14-331294.04.01.01</td>\n",
       "      <td>left</td>\n",
       "      <td>VISpm4</td>\n",
       "      <td>331294</td>\n",
       "      <td>Vip-IRES-Cre;Ai14-331294</td>\n",
       "      <td>M</td>\n",
       "      <td>P55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full</td>\n",
       "      <td>0.361015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601506492</td>\n",
       "      <td>PS0810_E1-50_S88</td>\n",
       "      <td>BT035</td>\n",
       "      <td>CS180626100018</td>\n",
       "      <td>Vip Gpc3 Slc18a3</td>\n",
       "      <td>Vip-MET-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601790961</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663.04.01.01</td>\n",
       "      <td>left</td>\n",
       "      <td>VISpm5</td>\n",
       "      <td>331663</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663</td>\n",
       "      <td>F</td>\n",
       "      <td>P57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full</td>\n",
       "      <td>0.621829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601790945</td>\n",
       "      <td>PS0817_E1-50_S19</td>\n",
       "      <td>BT036</td>\n",
       "      <td>CS180626100041</td>\n",
       "      <td>Sst Tac2 Myh4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601803754</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663.04.02.02</td>\n",
       "      <td>right</td>\n",
       "      <td>VISp2/3</td>\n",
       "      <td>331663</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663</td>\n",
       "      <td>F</td>\n",
       "      <td>P57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full</td>\n",
       "      <td>0.164240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601803733</td>\n",
       "      <td>PS0817_E1-50_S25</td>\n",
       "      <td>BT036</td>\n",
       "      <td>CS180626100035</td>\n",
       "      <td>Sst Calb2 Pdlim5</td>\n",
       "      <td>Sst-MET-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601808698</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663.04.02.03</td>\n",
       "      <td>right</td>\n",
       "      <td>VISp5</td>\n",
       "      <td>331663</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663</td>\n",
       "      <td>F</td>\n",
       "      <td>P57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full</td>\n",
       "      <td>0.483030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601808619</td>\n",
       "      <td>PS0817_E1-50_S26</td>\n",
       "      <td>BT036</td>\n",
       "      <td>CS180626100043</td>\n",
       "      <td>Sst Hpse Cbln4</td>\n",
       "      <td>Sst-MET-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601810307</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663.04.02.04</td>\n",
       "      <td>right</td>\n",
       "      <td>VISp2/3</td>\n",
       "      <td>331663</td>\n",
       "      <td>Sst-IRES-Cre;Ai14-331663</td>\n",
       "      <td>F</td>\n",
       "      <td>P57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>full</td>\n",
       "      <td>0.295656</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601810292</td>\n",
       "      <td>PS0817_E1-50_S27</td>\n",
       "      <td>BT036</td>\n",
       "      <td>CS180626100034</td>\n",
       "      <td>Sst Calb2 Necab1</td>\n",
       "      <td>Sst-MET-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      project  cell_specimen_id                 cell_specimen_name hemisphere  \\\n",
       "0  mIVSCC-MET         601506507  Vip-IRES-Cre;Ai14-331294.04.01.01       left   \n",
       "1  mIVSCC-MET         601790961  Sst-IRES-Cre;Ai14-331663.04.01.01       left   \n",
       "2  mIVSCC-MET         601803754  Sst-IRES-Cre;Ai14-331663.04.02.02      right   \n",
       "3  mIVSCC-MET         601808698  Sst-IRES-Cre;Ai14-331663.04.02.03      right   \n",
       "4  mIVSCC-MET         601810307  Sst-IRES-Cre;Ai14-331663.04.02.04      right   \n",
       "\n",
       "  structure  donor_id                donor_name biological_sex  age  \\\n",
       "0    VISpm4    331294  Vip-IRES-Cre;Ai14-331294              M  P55   \n",
       "1    VISpm5    331663  Sst-IRES-Cre;Ai14-331663              F  P57   \n",
       "2   VISp2/3    331663  Sst-IRES-Cre;Ai14-331663              F  P57   \n",
       "3     VISp5    331663  Sst-IRES-Cre;Ai14-331663              F  P57   \n",
       "4   VISp2/3    331663  Sst-IRES-Cre;Ai14-331663              F  P57   \n",
       "\n",
       "   ethnicity  ...  apical_dendrite_status neuron_reconstruction_type  \\\n",
       "0        NaN  ...                     NaN                       full   \n",
       "1        NaN  ...                     NaN                       full   \n",
       "2        NaN  ...                     NaN                       full   \n",
       "3        NaN  ...                     NaN                       full   \n",
       "4        NaN  ...                     NaN                       full   \n",
       "\n",
       "  cell_soma_normalized_depth depth_from_pia_um ephys_session_id  \\\n",
       "0                   0.361015               NaN        601506492   \n",
       "1                   0.621829               NaN        601790945   \n",
       "2                   0.164240               NaN        601803733   \n",
       "3                   0.483030               NaN        601808619   \n",
       "4                   0.295656               NaN        601810292   \n",
       "\n",
       "   transcriptomics_sample_id  transcriptomics_batch  T-type Accession ID  \\\n",
       "0           PS0810_E1-50_S88                  BT035       CS180626100018   \n",
       "1           PS0817_E1-50_S19                  BT036       CS180626100041   \n",
       "2           PS0817_E1-50_S25                  BT036       CS180626100035   \n",
       "3           PS0817_E1-50_S26                  BT036       CS180626100043   \n",
       "4           PS0817_E1-50_S27                  BT036       CS180626100034   \n",
       "\n",
       "       T-type Label MET-type Label  \n",
       "0  Vip Gpc3 Slc18a3      Vip-MET-2  \n",
       "1     Sst Tac2 Myh4            NaN  \n",
       "2  Sst Calb2 Pdlim5      Sst-MET-3  \n",
       "3    Sst Hpse Cbln4      Sst-MET-8  \n",
       "4  Sst Calb2 Necab1      Sst-MET-3  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.path.append(\"../downloads\")\n",
    "metadata = pd.read_csv(\"../downloads/20200711_patchseq_metadata_mouse.csv\")\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "File manifest\n",
    "The file manifest contains URLs for the different data files associated with the Patch-seq cells that are located in various archives."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_size</th>\n",
       "      <th>sha256_checksum</th>\n",
       "      <th>md5_checksum</th>\n",
       "      <th>technique</th>\n",
       "      <th>file_type</th>\n",
       "      <th>archive</th>\n",
       "      <th>archive_uri</th>\n",
       "      <th>project</th>\n",
       "      <th>cell_specimen_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>601506507</td>\n",
       "      <td>601506507.marker</td>\n",
       "      <td>268</td>\n",
       "      <td>305938ce75d55c3f43fbbe2d1d57e203139fa0c357bd73...</td>\n",
       "      <td>ecf2285ccbf5274af7d6a31a855ce950</td>\n",
       "      <td>morphology</td>\n",
       "      <td>marker_swc</td>\n",
       "      <td>BIL</td>\n",
       "      <td>ftp://download.brainlib.org:8811/biccn/zeng/ps...</td>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>601506507</td>\n",
       "      <td>601506507_raw.swc</td>\n",
       "      <td>160503</td>\n",
       "      <td>6c82f5f974e4f6d2336a3606d4955823edd3c0173da60d...</td>\n",
       "      <td>7cc7ccdd627ef83fe54fcbf0feac79c2</td>\n",
       "      <td>morphology</td>\n",
       "      <td>raw_swc</td>\n",
       "      <td>BIL</td>\n",
       "      <td>ftp://download.brainlib.org:8811/biccn/zeng/ps...</td>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>601506507</td>\n",
       "      <td>601506507_transformed.swc</td>\n",
       "      <td>271764</td>\n",
       "      <td>bad13a99c8f7f8a375e43109d80fd23d7028555a7dc42b...</td>\n",
       "      <td>f6a365d86ff65370af26439f0f16caaf</td>\n",
       "      <td>morphology</td>\n",
       "      <td>transformed_swc</td>\n",
       "      <td>BIL</td>\n",
       "      <td>ftp://download.brainlib.org:8811/biccn/zeng/ps...</td>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>601506492</td>\n",
       "      <td>sub-599387254_ses-601506492_icephys.nwb</td>\n",
       "      <td>17881801</td>\n",
       "      <td>54003c5b599fa8081938713e8c7c3b69d1883b31b3ada1...</td>\n",
       "      <td>d854a21668f5f2229024d8cf0975de54</td>\n",
       "      <td>intracellular_electrophysiology</td>\n",
       "      <td>nwb</td>\n",
       "      <td>DANDI</td>\n",
       "      <td>https://api.dandiarchive.org/api/assets/e734ba...</td>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>609481756</td>\n",
       "      <td>PS0810_E1-50_GCTCATGA-TCTCTCCG_R2.fastq.gz</td>\n",
       "      <td>48014090</td>\n",
       "      <td>a2954fc88cb34bc30b9bca9e9f1a0cff2c9b53cf9a7268...</td>\n",
       "      <td>fc9441d44f1fab3e5b31cbb2c57228b0</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>reverse_fastq</td>\n",
       "      <td>NEMO</td>\n",
       "      <td>http://data.nemoarchive.org/other/AIBS/AIBS_pa...</td>\n",
       "      <td>mIVSCC-MET</td>\n",
       "      <td>601506507.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     file_id                                   file_name  file_size  \\\n",
       "0  601506507                            601506507.marker        268   \n",
       "1  601506507                           601506507_raw.swc     160503   \n",
       "2  601506507                   601506507_transformed.swc     271764   \n",
       "3  601506492     sub-599387254_ses-601506492_icephys.nwb   17881801   \n",
       "4  609481756  PS0810_E1-50_GCTCATGA-TCTCTCCG_R2.fastq.gz   48014090   \n",
       "\n",
       "                                     sha256_checksum  \\\n",
       "0  305938ce75d55c3f43fbbe2d1d57e203139fa0c357bd73...   \n",
       "1  6c82f5f974e4f6d2336a3606d4955823edd3c0173da60d...   \n",
       "2  bad13a99c8f7f8a375e43109d80fd23d7028555a7dc42b...   \n",
       "3  54003c5b599fa8081938713e8c7c3b69d1883b31b3ada1...   \n",
       "4  a2954fc88cb34bc30b9bca9e9f1a0cff2c9b53cf9a7268...   \n",
       "\n",
       "                       md5_checksum                        technique  \\\n",
       "0  ecf2285ccbf5274af7d6a31a855ce950                       morphology   \n",
       "1  7cc7ccdd627ef83fe54fcbf0feac79c2                       morphology   \n",
       "2  f6a365d86ff65370af26439f0f16caaf                       morphology   \n",
       "3  d854a21668f5f2229024d8cf0975de54  intracellular_electrophysiology   \n",
       "4  fc9441d44f1fab3e5b31cbb2c57228b0                  transcriptomics   \n",
       "\n",
       "         file_type archive                                        archive_uri  \\\n",
       "0       marker_swc     BIL  ftp://download.brainlib.org:8811/biccn/zeng/ps...   \n",
       "1          raw_swc     BIL  ftp://download.brainlib.org:8811/biccn/zeng/ps...   \n",
       "2  transformed_swc     BIL  ftp://download.brainlib.org:8811/biccn/zeng/ps...   \n",
       "3              nwb   DANDI  https://api.dandiarchive.org/api/assets/e734ba...   \n",
       "4    reverse_fastq    NEMO  http://data.nemoarchive.org/other/AIBS/AIBS_pa...   \n",
       "\n",
       "      project  cell_specimen_id  \n",
       "0  mIVSCC-MET       601506507.0  \n",
       "1  mIVSCC-MET       601506507.0  \n",
       "2  mIVSCC-MET       601506507.0  \n",
       "3  mIVSCC-MET       601506507.0  \n",
       "4  mIVSCC-MET       601506507.0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change - file is actually an xcel\n",
    "file_manifest = pd.read_excel(\"../downloads/2021-09-13_mouse_file_manifest.xlsx\")\n",
    "file_manifest.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcriptomic data\n",
    "Here, we have already downloaded the (large) transcriptomic expression data file. In that file, expression is quantified as counts-per-million (cpm), and each row is a gene and each column is a cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = pd.read_csv(\n",
    "    \"../downloads/20200513_Mouse_PatchSeq_Release_cpm.v2.csv\",\n",
    "    index_col=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PS0810_E1-50_S88</th>\n",
       "      <th>PS0817_E1-50_S19</th>\n",
       "      <th>PS0817_E1-50_S25</th>\n",
       "      <th>PS0817_E1-50_S26</th>\n",
       "      <th>PS0817_E1-50_S27</th>\n",
       "      <th>PS0817_E1-50_S28</th>\n",
       "      <th>PS0817_E1-50_S46</th>\n",
       "      <th>PS0817_E1-50_S52</th>\n",
       "      <th>PS0830_E1-50_S17</th>\n",
       "      <th>PS0830_E1-50_S19</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-J39ZH_S561_E1-50</th>\n",
       "      <th>SM-J39ZH_S562_E1-50</th>\n",
       "      <th>SM-J39ZH_S565_E1-50</th>\n",
       "      <th>SM-J39ZH_S566_E1-50</th>\n",
       "      <th>SM-J39ZH_S571_E1-50</th>\n",
       "      <th>SM-J39ZH_S576_E1-50</th>\n",
       "      <th>SM-J3A1L_S592_E1-50</th>\n",
       "      <th>SM-J3A1L_S593_E1-50</th>\n",
       "      <th>SM-J3A1L_S603_E1-50</th>\n",
       "      <th>SM-J3A1L_S604_E1-50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>152.239614</td>\n",
       "      <td>165.312294</td>\n",
       "      <td>58.189895</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>551.71563</td>\n",
       "      <td>109.361330</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.223984</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>24.467996</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>51.366477</td>\n",
       "      <td>91.617703</td>\n",
       "      <td>82.706145</td>\n",
       "      <td>116.083588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69.658825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>57.335504</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>78.641855</td>\n",
       "      <td>...</td>\n",
       "      <td>52.858114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.620558</td>\n",
       "      <td>27.46475</td>\n",
       "      <td>40.197422</td>\n",
       "      <td>70.429453</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>111.937746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.662568</td>\n",
       "      <td>33.786897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.715418</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.747714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>31.342899</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>103.646061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s142</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s143</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s144</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s146</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s149</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45768 rows × 4435 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               PS0810_E1-50_S88  PS0817_E1-50_S19  PS0817_E1-50_S25  \\\n",
       "0610005C13Rik               0.0               0.0          0.000000   \n",
       "0610006L08Rik               0.0               0.0          0.000000   \n",
       "0610007P14Rik               0.0               0.0          0.000000   \n",
       "0610009B22Rik               0.0               0.0         69.658825   \n",
       "0610009E02Rik               0.0               0.0          0.000000   \n",
       "...                         ...               ...               ...   \n",
       "n-R5s142                    0.0               0.0          0.000000   \n",
       "n-R5s143                    0.0               0.0          0.000000   \n",
       "n-R5s144                    0.0               0.0          0.000000   \n",
       "n-R5s146                    0.0               0.0          0.000000   \n",
       "n-R5s149                    0.0               0.0          0.000000   \n",
       "\n",
       "               PS0817_E1-50_S26  PS0817_E1-50_S27  PS0817_E1-50_S28  \\\n",
       "0610005C13Rik          0.000000          0.000000          0.000000   \n",
       "0610006L08Rik          0.000000          0.000000          0.000000   \n",
       "0610007P14Rik        152.239614        165.312294         58.189895   \n",
       "0610009B22Rik          0.000000          0.000000          0.000000   \n",
       "0610009E02Rik          0.000000          0.000000          1.662568   \n",
       "...                         ...               ...               ...   \n",
       "n-R5s142               0.000000          0.000000          0.000000   \n",
       "n-R5s143               0.000000          0.000000          0.000000   \n",
       "n-R5s144               0.000000          0.000000          0.000000   \n",
       "n-R5s146               0.000000          0.000000          0.000000   \n",
       "n-R5s149               0.000000          0.000000          0.000000   \n",
       "\n",
       "               PS0817_E1-50_S46  PS0817_E1-50_S52  PS0830_E1-50_S17  \\\n",
       "0610005C13Rik          0.000000          0.000000           0.00000   \n",
       "0610006L08Rik          0.000000          0.000000           0.00000   \n",
       "0610007P14Rik          0.000000          0.000000         551.71563   \n",
       "0610009B22Rik          0.000000         57.335504           0.00000   \n",
       "0610009E02Rik         33.786897          0.000000           0.00000   \n",
       "...                         ...               ...               ...   \n",
       "n-R5s142               0.000000          0.000000           0.00000   \n",
       "n-R5s143               0.000000          0.000000           0.00000   \n",
       "n-R5s144               0.000000          0.000000           0.00000   \n",
       "n-R5s146               0.000000          0.000000           0.00000   \n",
       "n-R5s149               0.000000          0.000000           0.00000   \n",
       "\n",
       "               PS0830_E1-50_S19  ...  SM-J39ZH_S561_E1-50  \\\n",
       "0610005C13Rik          0.000000  ...             0.000000   \n",
       "0610006L08Rik          0.000000  ...             0.000000   \n",
       "0610007P14Rik        109.361330  ...             0.000000   \n",
       "0610009B22Rik         78.641855  ...            52.858114   \n",
       "0610009E02Rik          0.000000  ...             0.000000   \n",
       "...                         ...  ...                  ...   \n",
       "n-R5s142               0.000000  ...             0.000000   \n",
       "n-R5s143               0.000000  ...             0.000000   \n",
       "n-R5s144               0.000000  ...             0.000000   \n",
       "n-R5s146               0.000000  ...             0.000000   \n",
       "n-R5s149               0.000000  ...             0.000000   \n",
       "\n",
       "               SM-J39ZH_S562_E1-50  SM-J39ZH_S565_E1-50  SM-J39ZH_S566_E1-50  \\\n",
       "0610005C13Rik                  0.0             0.000000              0.00000   \n",
       "0610006L08Rik                  0.0             0.000000              0.00000   \n",
       "0610007P14Rik                  0.0            32.223984              0.00000   \n",
       "0610009B22Rik                  0.0            27.620558             27.46475   \n",
       "0610009E02Rik                  0.0            20.715418              0.00000   \n",
       "...                            ...                  ...                  ...   \n",
       "n-R5s142                       0.0             0.000000              0.00000   \n",
       "n-R5s143                       0.0             0.000000              0.00000   \n",
       "n-R5s144                       0.0             0.000000              0.00000   \n",
       "n-R5s146                       0.0             0.000000              0.00000   \n",
       "n-R5s149                       0.0             0.000000              0.00000   \n",
       "\n",
       "               SM-J39ZH_S571_E1-50  SM-J39ZH_S576_E1-50  SM-J3A1L_S592_E1-50  \\\n",
       "0610005C13Rik             0.000000             0.000000             0.000000   \n",
       "0610006L08Rik             0.000000             0.000000             0.000000   \n",
       "0610007P14Rik            24.467996             0.000000            51.366477   \n",
       "0610009B22Rik            40.197422            70.429453             0.000000   \n",
       "0610009E02Rik             1.747714             0.000000             0.000000   \n",
       "...                            ...                  ...                  ...   \n",
       "n-R5s142                  0.000000             0.000000             0.000000   \n",
       "n-R5s143                  0.000000             0.000000             0.000000   \n",
       "n-R5s144                  0.000000             0.000000             0.000000   \n",
       "n-R5s146                  0.000000             0.000000             0.000000   \n",
       "n-R5s149                  0.000000             0.000000             0.000000   \n",
       "\n",
       "               SM-J3A1L_S593_E1-50  SM-J3A1L_S603_E1-50  SM-J3A1L_S604_E1-50  \n",
       "0610005C13Rik             0.000000             0.000000             0.000000  \n",
       "0610006L08Rik             0.000000             0.000000             0.000000  \n",
       "0610007P14Rik            91.617703            82.706145           116.083588  \n",
       "0610009B22Rik             0.000000             0.000000           111.937746  \n",
       "0610009E02Rik            31.342899             0.000000           103.646061  \n",
       "...                            ...                  ...                  ...  \n",
       "n-R5s142                  0.000000             0.000000             0.000000  \n",
       "n-R5s143                  0.000000             0.000000             0.000000  \n",
       "n-R5s144                  0.000000             0.000000             0.000000  \n",
       "n-R5s146                  0.000000             0.000000             0.000000  \n",
       "n-R5s149                  0.000000             0.000000             0.000000  \n",
       "\n",
       "[45768 rows x 4435 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's create a 2D projection of the data using the transcriptomic data to see the structure of the different subclasses and types. We will transform the data, then use the UMAP algorithm to create a nonlinear embedding of a subset of the differentially expressed genes.\n",
    "\n",
    "the select_markers.csv was not available and so need to be generated. I am trying this method\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3801\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mc:\\anaconda3\\Lib\\site-packages\\pandas\\_libs\\index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\anaconda3\\Lib\\site-packages\\pandas\\_libs\\index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Gene'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 7\u001b[0m\n\u001b[0;32m      3\u001b[0m marker_genes_for_umap \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m../downloads/select_markers.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m, index_col\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[0;32m      5\u001b[0m \u001b[38;5;66;03m#need to modify select_markers so it is single column, unique and confirm exsist in gene_data see generate_select markers py\u001b[39;00m\n\u001b[0;32m      6\u001b[0m embedding \u001b[38;5;241m=\u001b[39m umap\u001b[38;5;241m.\u001b[39mUMAP(n_neighbors\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m25\u001b[39m)\u001b[38;5;241m.\u001b[39mfit_transform(\n\u001b[1;32m----> 7\u001b[0m     np\u001b[38;5;241m.\u001b[39mlog2(gene_data\u001b[38;5;241m.\u001b[39mloc[marker_genes_for_umap[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGene\u001b[39m\u001b[38;5;124m\"\u001b[39m], :]\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mT \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m      8\u001b[0m )\n",
      "File \u001b[1;32mc:\\anaconda3\\Lib\\site-packages\\pandas\\core\\frame.py:3807\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3805\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   3806\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 3807\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(key)\n\u001b[0;32m   3808\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   3809\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m-> 3804\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3805\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3806\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3807\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3808\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3809\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Gene'"
     ]
    }
   ],
   "source": [
    "import umap\n",
    "\n",
    "marker_genes_for_umap = pd.read_csv(\"../downloads/select_markers.csv\", index_col=0)\n",
    "\n",
    "#need to modify select_markers so it is single column, unique and confirm exsist in gene_data see generate_select markers py\n",
    "embedding = umap.UMAP(n_neighbors=25).fit_transform(\n",
    "    np.log2(gene_data.loc[marker_genes_for_umap[\"Gene\"], :].values.T + 1)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 8))\n",
    "plt.scatter(*embedding.T, s=1, edgecolor=\"none\")\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see where specific types fall in this embedding by using the metadata to identify specific cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the cells from a particular t-type (Lamp5 Plch2 Dock5)\n",
    "my_ttype_metadata = metadata.loc[metadata[\"T-type Label\"] == \"Lamp5 Plch2 Dock5\", :]\n",
    "\n",
    "my_ttype_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_ttype_mask = gene_data.columns.isin(my_ttype_metadata[\"transcriptomics_sample_id\"].tolist())\n",
    "\n",
    "plt.figure(figsize=(8, 8))\n",
    "plt.scatter(*embedding.T, s=1, edgecolor=\"none\")\n",
    "plt.scatter(*embedding[my_ttype_mask, :].T, s=2, edgecolor=\"none\")\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also look at how expression of a particular gene varies across types in this embedding."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also look at how expression of a particular gene varies across types in this embedding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 8))\n",
    "plt.scatter(\n",
    "    *embedding.T,\n",
    "    s=1,\n",
    "    c=gene_data.loc[\"Npy\", :].values,\n",
    "    vmin=0,\n",
    "    vmax=5e3,\n",
    "    cmap=\"viridis\",\n",
    "    edgecolor=\"none\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding cells with electrophysiology and morphology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_ttype_with_recon_metadata = metadata.loc[\n",
    "    (metadata[\"T-type Label\"] == \"Lamp5 Plch2 Dock5\") &\n",
    "    (metadata[\"neuron_reconstruction_type\"].isin([\"full\"])),\n",
    "    :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_ttype_with_recon_metadata.iloc[0, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Electrophysiology\n",
    "Let's get the NWB file that has the electrophysiology data for this cell and use the IPFX library to process it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_specimen_id = my_ttype_with_recon_metadata.iloc[0, :][\"cell_specimen_id\"]\n",
    "\n",
    "nwb_urls = file_manifest.loc[\n",
    "    (file_manifest[\"cell_specimen_id\"] == float(my_specimen_id)) &\n",
    "    (file_manifest[\"file_type\"] == \"nwb\"),\n",
    "    :\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nwb_urls[\"archive_uri\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!dandi download https://api.dandiarchive.org/api/assets/5a0d8719-3b7c-41f7-b235-3640d3f242e7/download/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nwb_path = nwb_urls[\"file_name\"].values[0]\n",
    "nwb_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ipfx.dataset.create import create_ephys_data_set\n",
    "from ipfx.data_set_features import extract_data_set_features\n",
    "from ipfx.utilities import drop_failed_sweeps\n",
    "\n",
    "data_set = create_ephys_data_set(nwb_file=nwb_path)\n",
    "drop_failed_sweeps(data_set)\n",
    "cell_features, sweep_features, cell_record, sweep_records, _, _ = \\\n",
    "    extract_data_set_features(data_set, subthresh_min_amp=-100.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_features.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_features[\"long_squares\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_features[\"long_squares\"][\"rheobase_sweep\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_features[\"long_squares\"][\"rheobase_sweep\"][\"latency\"]\n",
    "df=cell_features[\"long_squares\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "swp = data_set.sweep(cell_features[\"long_squares\"][\"rheobase_sweep\"][\"sweep_number\"])\n",
    "\n",
    "swp = data_set.sweep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "plt.plot(swp.t, swp.v)\n",
    "plt.xlabel(\"time (s)\", fontsize=16)\n",
    "plt.ylabel(\"membrane potential (mV)\", fontsize=16)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Morphology\n",
    "Now we'll get the SWC file that has the morphological reconstruction of this cell and use the neuron_morphology library to process it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "swc_urls = file_manifest.loc[\n",
    "    (file_manifest[\"cell_specimen_id\"] == float(my_specimen_id)) &\n",
    "    (file_manifest[\"file_type\"] == \"transformed_swc\"),\n",
    "    :\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "swc_urls[\"archive_uri\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget ftp://download.brainlib.org:8811/biccn/zeng/pseq/morph/200526/645169930_transformed.swc"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
